File: Top 50 Coding Interview Questions/41 Write a Program to Add two Fractions/AddtwoFractions.cpp
#include <iostream>
using namespace std;

// Function to calculate the GCD (Greatest Common Divisor) using the Euclidean algorithm
int findGCD(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

int main() {
    int num1, den1, num2, den2;

    // Input first fraction
    cout << "Enter numerator and denominator of the first fraction: ";
    cin >> num1 >> den1;

    // Input second fraction
    cout << "Enter numerator and denominator of the second fraction: ";
    cin >> num2 >> den2;

    // Calculate LCM (Least Common Multiple) of the denominators
    int gcdDen = findGCD(den1, den2);
    int lcm = (den1 * den2) / gcdDen;

    // Calculate the sum of fractions
    int numeratorSum = (num1 * (lcm / den1)) + (num2 * (lcm / den2));

    // Simplify the result by finding GCD of the numerator and denominator
    int gcdResult = findGCD(numeratorSum, lcm);
    int simplifiedNumerator = numeratorSum / gcdResult;
    int simplifiedDenominator = lcm / gcdResult;

    // Output the result
    cout << num1 << "/" << den1 << " + " << num2 << "/" << den2
         << " = " << simplifiedNumerator << "/" << simplifiedDenominator << endl;

    return 0;
}

File: Top 50 Coding Interview Questions/41 Write a Program to Add two Fractions/AddtwoFractions.py
def findGCD(n1, n2):
    """
    Function to calculate the Greatest Common Divisor (GCD) of two numbers
    using a simple iteration method.
    """
    gcd = 0
    # Loop from 1 to the minimum of n1 and n2 (inclusive)
    for i in range(1, int(min(n1, n2)) + 1):
        # Check if i is a divisor of both n1 and n2
        if n1 % i == 0 and n2 % i == 0:
            gcd = i
    return gcd

# Input first fraction
num1, den1 = map(int, input("Enter numerator and denominator of the first fraction (space-separated): ").split())

# Input second fraction
num2, den2 = map(int, input("Enter numerator and denominator of the second fraction (space-separated): ").split())

# Calculate LCM (Least Common Multiple) of the denominators
lcm = (den1 * den2) // findGCD(den1, den2)

# Calculate the numerator of the sum of the fractions
sum_num = (num1 * lcm // den1) + (num2 * lcm // den2)

# Simplify the resulting fraction by dividing numerator and denominator by their GCD
gcd_result = findGCD(sum_num, lcm)
simplified_num = sum_num // gcd_result
simplified_den = lcm // gcd_result

# Output the result
print(f"{num1}/{den1} + {num2}/{den2} = {simplified_num}/{simplified_den}")
